misk/src/test/kotlin/misk/tasks/RepeatedTaskQueueTest.kt

```diff
@@ -0,0 +1,486 @@
+      lock.withLock {
+        clock.add(Duration.ofSeconds(35))
+        taskCompleted.signalAll()
+        throw IllegalArgumentException("this failed")
+      }
+    }
+
+    // Repeat the task 5 times to keep expanding the backoff
+    for (i in 0 until 5) {
+      lock.withLock {
+        pendingTasks.release(1)
+        assertThat(taskCompleted.await(5, TimeUnit.SECONDS)).isTrue()
+      }
+
+      // Wait for the task to schedule its follow up
+      waitForNextPendingTask()
+    }
+
+    // Next task should be scheduled at the failure backoff
+    val nextTask = waitForNextPendingTask()
+    assertThat(nextTask.getDelay(TimeUnit.SECONDS)).isEqualTo(60L)
+  }
+
+  @Test fun honorsFailureBackoffOnError() {
+    val lock = ReentrantLock()
+    val taskCompleted = lock.newCondition()
+
+    // Create a task that wakes up the main thread on execution, and then fails
+    taskQueue.scheduleWithBackoff(
+        timeBetweenRuns = Duration.ofSeconds(2),
+        failureBackoff = ExponentialBackoff(Duration.ofSeconds(5), Duration.ofSeconds(60)),
+        noWorkBackoff = ExponentialBackoff(Duration.ofSeconds(10), Duration.ofSeconds(100))) {
+      lock.withLock {
+        clock.add(Duration.ofSeconds(35))
+        taskCompleted.signalAll()
+        Status.FAILED
+      }
+    }
+
+    // Repeat the task 5 times to keep expanding the backoff
+    for (i in 0 until 5) {
+      lock.withLock {
+        pendingTasks.release(1)
+        assertThat(taskCompleted.await(5, TimeUnit.SECONDS)).isTrue()
+      }
+
+      // Wait for the task to schedule its follow up
+      waitForNextPendingTask()
+    }
+
+    // Next task should be scheduled at the failure backoff
+    val nextTask = waitForNextPendingTask()
+    assertThat(nextTask.getDelay(TimeUnit.SECONDS)).isEqualTo(60L)
+  }
+
+  @Test fun honorsNoWorkBackoff() {
+    val lock = ReentrantLock()
+    val taskCompleted = lock.newCondition()
+
+    // Create a task that wakes up the main thread on execution, and then returns with no work
+    taskQueue.scheduleWithBackoff(
+        timeBetweenRuns = Duration.ofSeconds(2),
+        failureBackoff = ExponentialBackoff(Duration.ofSeconds(5), Duration.ofSeconds(60)),
+        noWorkBackoff = ExponentialBackoff(Duration.ofSeconds(10), Duration.ofSeconds(100))) {
+      lock.withLock {
+        clock.add(Duration.ofSeconds(35))
+        taskCompleted.signalAll()
+        Status.NO_WORK
+      }
+    }
+
+    // Repeat the task 5 times to keep expanding the backoff
+    for (i in 0 until 5) {
+      lock.withLock {
+        pendingTasks.release(1)
+        assertThat(taskCompleted.await(5, TimeUnit.SECONDS)).isTrue()
+      }
+
+      // Wait for the task to schedule its follow up
+      waitForNextPendingTask()
+    }
+
+    // Next task should be scheduled at the no-work backoff
+    val nextTask = waitForNextPendingTask()
+    assertThat(nextTask.getDelay(TimeUnit.SECONDS)).isEqualTo(100L)
+  }
+
+  @Test fun resetFailureBackoffOnSuccess() {
```

Comment:
/nit could these resetXonY tets be DRYed up a bit? Seems like they follow a similar pattern of setup -> perform 5 times in state X -> perform once in state Y -> perform once more in state X -> confirm final backoff